File: rxhttp/src/main/java/rxhttp/wrapper/parse/SmartParser.java
package rxhttp.wrapper.parse;

import java.io.IOException;
import java.lang.reflect.Type;

import okhttp3.Response;
import rxhttp.wrapper.utils.Converter;
import rxhttp.wrapper.utils.TypeUtil;

/**
 * Convert [Response] to [T]
 * User: ljx
 * Date: 2018/10/23
 * Time: 13:49
 */
public class SmartParser<T> extends TypeParser<T> {

    protected SmartParser() {
    }

    public SmartParser(Type type) {
        super(type);
    }

    @Override
    public T onParse(Response response) throws IOException {
        return Converter.convert(response, types[0]);
    }

    @SuppressWarnings({"unchecked", "rawtypes"})
    public static <T> Parser<T> wrap(Type type) {
        Type actualType = TypeUtil.getActualType(type);
        if (actualType == null) actualType = type;
        SmartParser smartParser = new SmartParser(actualType);
        return actualType == type ? smartParser : new OkResponseParser(smartParser);
    }
}

File: rxhttp/src/main/java/rxhttp/wrapper/parse/SmartParser.kt
package rxhttp.wrapper.parse

import okhttp3.Response
import rxhttp.wrapper.utils.TypeUtil
import rxhttp.wrapper.utils.convert
import java.io.IOException
import java.lang.reflect.Type

/**
 * Convert [Response] to [T]
 * User: ljx
 * Date: 2018/10/23
 * Time: 13:49
 */
open class SmartParser<T> : TypeParser<T> {
    protected constructor() : super()
    constructor(type: Type) : super(type)

    @Throws(IOException::class)
    override fun onParse(response: Response): T = response.convert(types[0])

    companion object {
        @Suppress("UNCHECKED_CAST")
        @JvmStatic
        fun <T> wrap(type: Type): Parser<T> {
            val actualType = TypeUtil.getActualType(type) ?: type
            val parser = SmartParser<Any>(actualType)
            val actualParser = if (actualType == type) parser else OkResponseParser(parser)
            return actualParser as Parser<T>
        }
    }
}